test/xbgpu/test_engine.py

```diff
@@ -1452,3 +1555,153 @@ def get_stream_status(stream_name: str) -> bool:
+                    beam_outputs
+                )
+            capture_stop_beams = []
+            for stopped_beam in stopped_beams:
+                capture_stop_beams.append(("capture-stop", stopped_beam.name))
+            # NOTE: `get_free_chunk` is given `capture_stop_index` + 1 because
+            # `patch_method` counts from 1 instead of 0.
+            self._patch_method(
+                monkeypatch,
+                bsend.BSend,
+                "get_free_chunk",
+                capture_stop_chunk_index + 1,
+                client,
+                capture_stop_beams,
+            )
+
+        timestamp_step = n_samples_between_spectra * n_spectra_per_heap
+        corrprod_results, beam_results, _, _ = await self._send_data(
+            mock_recv_streams,
+            mock_send_stream,
+            corrprod_outputs,
+            beam_outputs,
+            heap_factory=lambda batch_index, present: self._default_heap_factory(
+                timestamp_step=timestamp_step,
+                n_ants=n_ants,
+                n_channels_per_substream=n_channels_per_substream,
+                n_spectra_per_heap=n_spectra_per_heap,
+                frequency=frequency,
+                batch_index=batch_index,
+                present=present,
+            ),
+            present=np.ones((n_batches, n_ants), bool),
+            timestamp_step=timestamp_step,
+            n_channels_per_substream=n_channels_per_substream,
+            frequency=frequency,
+            n_spectra_per_heap=n_spectra_per_heap,
+            corrprod_capture_stop_accum_indices=corrprod_capture_stop_accum_indices,
+            beam_capture_stop_heap_indices=beam_capture_stop_heap_indices,
+        )
+
+        # NOTE: During receiving beam data, only data up to `beam_capture_stop_index`
+        # is stored. It is also verified that there is no more data in the stream.
+        # This is largely a sanity check that there is non-zero data for heaps that
+        # were transmitted before the `?capture-stop` request.
+        for beam_output in beam_outputs:
+            np.testing.assert_equal(beam_results[beam_output.name] != 0, True)
+
+        # NOTE: The X-engine output is entirely real (no imag component).
+        # The results buffer has shape (n_accumulations, n_channels_per_substream,
+        # n_baselines, COMPLEX). As a result, the final dimension only has data
+        # populated for the 'real' (first) index.
+        for corrprod_output, corrprod_capture_stop_accum_index in zip(
+            corrprod_outputs, corrprod_capture_stop_accum_indices
+        ):
+            if corrprod_capture_stop_accum_index is not None:
+                if corrprod_capture_stop_accum_index == 0:
+                    # The XPipeline never transmitted any accumulations
+                    # Make sure the ndarray is empty
+                    assert corrprod_results[corrprod_output.name].size == 0
```

Comment:
Is this code even reachable, given that you've explicitly set `capture_stop_accum_index` to 3?

Reply:
Fair enough. I'll remove.